doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+alternatives of the co-product and $i_j \hookrightarrow \tau_j$ maps the index
+$i_j$ to the corresponding type $\tau_j$. We use $\langle \ldots \rangle$ to
+indicate the labeling is now disjunctive and $i_k \cdot e$ as the injection of
+$e$ into the sum type using label $i_k$. Note the symmetry with complementary
+rule for projection out of a record: the only difference is that the (now
+disjunctive) row (resp. (now injected) expression) is swapped from premise to
+conclusion (resp. from conclusion to premise).
+
+This rule tells us that, when an expression $e$ with a type $t_k$ is derivable
+from a context, we can include it as an alternative in our sum type by injecting
+it with the label $i_k$, giving an element of our sum type. If we're given a
+thing that has a type allowed by our alternatives, it can included among our
+alternatives.
+
+The proposed rule in our system has been formed by seeking the same symmetry w/r/t
+projection out from a product:
+
+$$
+\frac
+{ \Gamma \vdash e \colon (t, c) \quad \Gamma \vdash \`l\` \colon str \quad fresh(s) }
+{ \Gamma \vdash \ inj(\`l\`, e) \ \colon (s, c \land s \sim \\{ \ l \colon t | tail\_s \ \\}) }
+$$
+
+I here retain our current practice of using quint string literals along with an
+exotic operator, tho in the final section I recommend we abandon this course.
+
+Comparing this with our current rule for projecting out of records, we see the
+same symmetry: the (now disjunctive) row type is now synthesized instead of
+being taken from the context.
+
+### Introducing products and eliminating sums
+
+Forming expressions of product types by backing them into records and
+eliminating expressions of sum types by case analysis exhibit the same duality,
+tho they are a bit more complex.
+
+#### Packing expressions into records
+
+Here is a concrete example of forming a record using our current syntax:
+
+```quint
+val n : int = 1
+val s : str = "one"
+val ex : {a : int, b : str} = {a : n, b : s}
+// Or, using our exotic Rec operator, which is currently the normal form
+val ex_ : {a : int, b : str} = Rec("a", n, "b", s)
+```
+
+A textbook introduction rule for finite products is given as
+
+$$ 
+\frac
+{ \Gamma \vdash e_1 \colon \tau_1 \quad \ldots \quad \Gamma \vdash e_n \colon \tau_n }
+{ \Gamma \vdash \\{ i_1 \hookrightarrow e_1, \ldots, i_n \hookrightarrow e_n \\} \colon \\{ i_1 \hookrightarrow \tau_1, \ldots, i_n \hookrightarrow \tau_n \\} }
+$$
+
+This tells us that for any expressions $e_1 : \tau_1, \ldots, e_n : \tau_n$
+derivable from our context we can form a product that indexes those
+$n$ expressions by $i_1, \ldots, i_n$ distinct labels, and packages all data
+together in an expression of type $\\{ i_1 \hookrightarrow \tau_1, \ldots,
+i_n \hookrightarrow \tau_n \\}$. If we're given all the things of the needed
+types, we can conjoint them all together into one compound package.
+
+The following rule describes our current implementation:
```

Comment:
This does not render in VSCode, and it renders strangely on Github. I must admit that reading complex LaTeX in markdown is quite hard. What I did not notice in this rule is a row variable. Maybe I just missed it.

Reply:
Sorry about that! I also had trouble with the rendering of these.

I can convert them to ascii if you think that would better?

Reply:
I've ported to org-mode and will include a generated pdf.